### PR TITLE
Simplify a few functions in rustc_data_structures

### PR DESCRIPTION
**Pull request by ljedrz**
- drop `try!()` where it's superfluous
- change `try!()` to `?`
- squash a `push` with `push_str`
- refactor a push loop into an iterator

### CODE REVIEW

**Comment by ljedrz** [comment]
Where's the CI?

File: src/librustc_data_structures/small_vec.rs [hunk 1]
```diff
@@ -209,11 +209,7 @@ impl<A> Decodable for SmallVec<A>
           A::Element: Decodable {
     fn decode<D: Decoder>(d: &mut D) -> Result<SmallVec<A>, D::Error> {
         d.read_seq(|d, len| {
-            let mut vec = SmallVec::with_capacity(len);
```
Comment (Xanewok) [inline src/librustc_data_structures/small_vec.rs]: I have a hunch that this was deliberately done for optimization

cc @nnethercote

Comment (nnethercote) [reply src/librustc_data_structures/small_vec.rs]: I don't have a good sense of how `collect()` works. Will it use `size_hint` from the range to pre-allocate an appropriate capacity?

Comment (Mark-Simulacrum) [reply src/librustc_data_structures/small_vec.rs]: It should, yes.

**Comment by kennytm** [comment]
(Re-triggering CI)

**Pull request merged**